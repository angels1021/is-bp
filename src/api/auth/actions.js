/*
 * LoginActions
 *
 * redux actions to describe login/logout actions
 *
 * */
import { createAction } from 'redux-actions';
import { createFormAction } from 'redux-form-saga';
import {
  LOGIN_REQUEST,
  LOGOUT_REQUEST,
  AUTH_REQUEST,
  AUTH_ERROR,
  AUTH_SUCCESS,
  AUTH_SET
} from './constants';

/**
 * Tells the app we want to login a user
 * @param  {object} formData          The formData we're sending for login
 * @param  {string} formData.username The username of the user to login
 * @param  {string} formData.password The password of the user to login
 * @param  {string} formData.location The location to navigate to after login
 * @param  {string} formData.code     The card code of the user to login (optional)
 *
 * @return {object} action object - { type: LOGIN_REQUEST, payload: formData{} }
 */
export const login = createAction(LOGIN_REQUEST, (formData) => formData.toObject()); // (formData) => formData

/**
 * Tells the app we want to log out a user
 *
 * @param {string/number} userId - the user to logout;
 *
 * @return {object} action object - { type: LOGOUT_REQUEST, payload: userId }
 */
export const logout = createAction(LOGOUT_REQUEST); // (userId) => userId

/**
 * Sets the 'error' state to error received (both login and logout use the same reducer)
 * @see http://redux-form.com/6.5.0/docs/api/SubmissionError.md/
 * for SubmissionError
 *
 * @param {object} Error/SubmissionError The error received when trying to make a request.
 *
 * @return {object} action object - { type: AUTH_ERROR, payload: error }
 */
export const authError = createAction(AUTH_ERROR); // (SubmissionError) => SubmissionError
// export const authError = (error) => ({
//   type: AUTH_ERROR,
//   payload: error
// }); // (error) => error

/**
 * Sets the 'pending' state to requested action (both login and logout use the same reducer)
 *
 * @param {string} type The action type making the request
 *
 * @return {object} action object - { type: AUTH_REQUEST, payload: LOGIN_REQUEST }
 */
export const authRequest = createAction(AUTH_REQUEST); // (type) => type

/**
 * Sets the 'user' state to requested response (both login and logout use the same reducer)
 *
 * @param {object} [user={}] The new user object to set. on logout it's unnecessary
 * @param {string} user.id The user id
 * @param {string} user.uuid The user's uuid
 * @param {string} user.role The user's authorized role
 * @param {string} user.firstName The user's first name
 * @param {string} user.lastName The user's last name
 *
 * @return {object} action object - { type: AUTH_SUCCESS, payload: user{} }
 */
export const authSuccess = createAction(AUTH_SUCCESS, (user) => user || {});

/**
 * Sets the authentication state of the application
 *
 * @param  {boolean} isLoggedIn
 */
export const authSet = createAction(AUTH_SET, (isLoggedIn) => Boolean(isLoggedIn));

export const loginAction = createFormAction(login, [AUTH_SUCCESS, AUTH_ERROR]);
