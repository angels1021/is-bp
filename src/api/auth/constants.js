/**
 * auth saga constants.
 */

export const LOGIN_REQUEST = 'api/auth/LOGIN_REQUEST';
export const LOGOUT_REQUEST = 'api/auth/LOGOUT_REQUEST';
export const AUTH_REQUEST = 'api/auth/AUTH_REQUEST';
export const AUTH_ERROR = 'api/auth/AUTH_ERROR';
export const AUTH_SUCCESS = 'api/auth/AUTH_SUCCESS';
export const AUTH_SET = 'api/auth/AUTH_SET';

export const initialState = {
  username: '',
  password: '',
  location: '/',
  code: ''
};
