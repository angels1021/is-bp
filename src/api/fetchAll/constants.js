/**
 * fetchAll api constants
 */

export const FETCH_ALL = 'api/fetchAll/FETCH_ALL';
