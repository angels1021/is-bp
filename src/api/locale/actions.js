/**
 * Locale api actions
 */
import {
  CHANGE_LOCALE_REQUEST,
  CHANGE_LOCALE_SUCCESS,
  CHANGE_LOCALE_FAIL,
  CHANGE_LOCALE_CANCEL
} from './constants';

/**
 * changeLocale - action creator
 *
 * @param {string} locale - two character language code, ie 'en'.
 * @param {object[]} request - request translations for the modules specified - passed to api/translations.
 * @property {string} request[].module - ie 'app'.
 * @property {string} request[].page ie  -'common'.
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocale = (locale, request, moduleId) => ({
  type: CHANGE_LOCALE_REQUEST,
  payload: { locale, request, moduleId }
});

/**
 * changeLocaleSuccess - action creator
 *
 * @param {string} locale - two character language code, ie 'en'.
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleSuccess = (locale, moduleId) => ({
  type: CHANGE_LOCALE_SUCCESS,
  payload: { locale, moduleId }
});

/**
 * changeLocaleFail - action creator
 *
 * @param {object} error - the response error
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleFail = (error, moduleId) => ({
  type: CHANGE_LOCALE_FAIL,
  payload: { error, moduleId }
});

/**
 * changeLocaleCancel - action creator
 *
 * @param {string} moduleId - the caller module who's reducer should catch this actions
 */
export const changeLocaleCancel = (moduleId) => ({
  type: CHANGE_LOCALE_CANCEL,
  payload: { moduleId }
});

