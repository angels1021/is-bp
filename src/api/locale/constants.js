/**
 * Locale api shared constants.
 */

export const CHANGE_LOCALE_REQUEST = 'api/locale/CHANGE_LOCALE_REQUEST';
export const CHANGE_LOCALE_SUCCESS = 'api/locale/CHANGE_LOCALE_SUCCESS';
export const CHANGE_LOCALE_FAIL = 'api/locale/CHANGE_LOCALE_FAIL';
export const CHANGE_LOCALE_CANCEL = 'api/locale/CHANGE_LOCALE_CANCEL';

