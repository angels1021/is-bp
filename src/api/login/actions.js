/*
* LoginActions
*
* redux actions to describe login/register actions
*
* */
import {
  SET_AUTH,
  REQUEST_PENDING,
  LOGIN_REQUEST,
  LOGOUT,
  REQUEST_ERROR,
  REQUEST_SUCCESS,
  CLEAR_ERROR
} from './constants';

/**
* requestSuccess - set the user received from login
*
* @param {type} user Description
* @param  {string} user.id The user id
* @param  {string} user.uuid The user's uuid
* @param  {string} user.role The user's authorized role
* @param  {string} user.firstName The user's first name
* @param  {string} user.lastName The user's last name
*
* @returns {type} Description
*/
export const requestSuccess = (user) => ({
  type: REQUEST_SUCCESS,
  user
});

/**
 * Sets the authentication state of the application
 *
 * @param  {boolean} newAuthState True means a user is logged in, false means no user is logged in
 */
export const setAuth = (newAuthState) => ({
  type: SET_AUTH,
  newAuthState: Boolean(newAuthState)
});

/**
 * Sets the `currentlySending` state, which displays a loading indicator during requests
 *
 * @param  {boolean} busy True means we're currently sending a request, False means we're idle
 */
export const sendingRequest = (pending) => ({
  type: REQUEST_PENDING,
  pending: Boolean(pending)
});

/**
 * Tells the app we want to login a user
 * @param  {object} data          The data we're sending for login
 * @param  {string} data.username The username of the user to login
 * @param  {string} data.password The password of the user to login
 * @param  {string} data.location The location to navigate to after login
 * @param  {string} data.code     The card code of the user to login (optional)
 */
export const loginRequest = (data) => ({
  type: LOGIN_REQUEST,
  data
});

/**
 * Tells the app we want to log out a user
 */
export const logoutRequest = () => ({
  type: LOGOUT
});


/**
 * Sets the 'error' state to error received
 *
 * @param {object} error The error received when trying to make arequest
 */
export const requestError = (error) => ({
  type: REQUEST_ERROR,
  error
});

/**
 * Sets the 'error' state as empty
 */
export const clearError = () => ({
  type: CLEAR_ERROR
});
