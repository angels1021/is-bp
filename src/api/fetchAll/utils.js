/**
 * Helper Generators not exclusive to fetchAll saga.
 */

import { fork, take, put, select, race, call } from 'redux-saga/effects';
import { MESSAGES_SUCCESS, MESSAGES_FAIL } from 'api/translations/constants';
import { requestMessages } from 'api/translations/actions';
import { selectRequestLoading, selectRequestErrors } from 'common/containers/App/selectors';

/**
 * fetch resources where their success/fail event are important to page load
 * for resources where it doesn't matter, call fork(put, action) instead.
 *
 * @param {object} action - redux action object
 * @param {string} takeSuccess - success constant to listen for
 * @param {string} takeFail - fail constant to listen for
 * @returns {boolean}
 */
export function* fetchResource(action, takeSuccess, takeFail) {
  yield put(action);
  const { fail, success } = yield race({
    success: take(takeSuccess),
    fail: take(takeFail)
  });
  if (fail) {
    throw fail.payload.error;
  }
  return success;
}

/**
 * high level call to fetchResource, specific to translations
 *
 * @param {function} selectLocale - selector function to get current module's locale
 * @param {object[]} request - request translations for the modules specified.
 * @property {string} request[].module - ie 'app'.
 * @property {string} request[].page ie  -'common'.
 */
export function* fetchMessages(selectLocale, request) {
  // prep messages
  const locale = yield select(selectLocale);
  // call messages
  yield fork(
    fetchResource,
    requestMessages(locale, request),
    MESSAGES_SUCCESS,
    MESSAGES_FAIL
  );
}

/**
 * helper generator for waitForOther
 *
 * @param {string} requestId         - requestId to track in global async requests
 * @param {function} loadingSelector - selector to check if the request is still loading
 * @param {function} errorSelector   - selector to check if the request has failed
 * @param {function} resolveSelector - selector function that must resolve to true before starting calls
 * @returns {{isLoading: boolean, resolved: boolean}}
 */
export function* checkSelectors({ requestId, loadingSelector, errorSelector, resolveSelector }) {
  const checkError = yield select(errorSelector);
  const hasError = checkError(requestId);
  // throw to fail child page waiting for a parent that failed
  if (hasError) throw hasError;
  const isLoading = yield select(loadingSelector);
  const resolved = yield select(resolveSelector);
  return { isLoading: isLoading(requestId), resolved };
}

/**
 * helper generator for callFetchSaga
 * wait for another module or page to finish loading before moving on
 *
 * @param {string} requestId          - requestId to track in global async requests
 * @param {function} [resolveSelector] - selector function that must resolve to true before starting calls
 */
export function* waitForOther(requestId, resolveSelector = () => true) {
  const loadingSelector = selectRequestLoading();
  const errorSelector = selectRequestErrors();
  const selectorOpts = { requestId, loadingSelector, errorSelector, resolveSelector };
  let { isLoading, resolved } = yield call(checkSelectors, selectorOpts);
  while (isLoading || !resolved) {
    const response = yield call(checkSelectors, selectorOpts);
    isLoading = response.isLoading;
    resolved = response.resolved;
  }
}

export const createRequestId = (moduleId) => `${moduleId}/FETCH_ALL`;
