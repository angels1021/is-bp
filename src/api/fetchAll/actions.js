/**
 * fetchAll api actions
 */
import { FETCH_ALL } from './constants';
import { createRequestId } from './utils';
/**
 * action creator for a fetchAll request.
 * expects an options object
 *
 * @param {string} requestId           - unique page/module id to track in global async reducer
 * @param {generator} fetchGenerator   - generator to call when ready to fetch resources
 * @param {string} [parentId]          - unique page/module id to wait for before starting calls
 * @param {function} [resolveSelector] - selector function that must resolve to true before starting calls
 */
export const fetchAllCreateSaga = ({ requestId, fetchGenerator, parentId, resolveSelector }) => ({
  type: FETCH_ALL,
  payload: {
    requestId: createRequestId(requestId),
    parentId: createRequestId(parentId),
    fetchGenerator,
    resolveSelector
  }
});

/**
 * actions creator for fetchAll events
 * should be listened to in individual module/page sagas
 *
 * @param moduleId
 */
export const fetchAll = (moduleId) => ({
  type: createRequestId(moduleId),
  payload: { id: moduleId }
});

