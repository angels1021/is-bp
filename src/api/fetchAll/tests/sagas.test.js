/**
 * Test fetchAll sagas
 */
/* eslint-disable redux-saga/yield-effects */
import { createMockTask } from 'redux-saga/utils';
import { takeEvery, takeLatest, fork, call, take, put, cancel } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { asyncSuccess, asyncFail, asyncRequest } from 'common/containers/App/actions';
import { waitForOther, createRequestId } from '../utils';
import { FETCH_ALL } from '../constants';
import {
  fetchWatcher,
  fetchFlow,
  watchFetchSaga,
  callFetchSaga
} from '../sagas';

describe('API fetchAll sagas', () => {
  const payload = {
    requestId: createRequestId('Clark'),
    parentId: createRequestId('Jonathan'),
    fetchGenerator: () => true,
    resolveSelector: () => true
  };
  describe('fetchWatcher', () => {
    it('should listen for any FETCH_ALL action ans start a fetchFlow', () => {
      // arrange
      const Generator = fetchWatcher();
      const expectedResult = fork(takeEvery, FETCH_ALL, fetchFlow);
      // act
      const result = Generator.next().value;
      // assert
      expect(result).toEqual(expectedResult);
    });
  });

  describe('fetchFlow', () => {
    it('should error if options.requestId or options.fetchGenerator were passed', () => {
      // assert
      expect(fetchFlow({})).toThrowErrorMatchingSnapshot();
    });
    const Generator = fetchFlow({ payload });
    const watcher = createMockTask();
    it('should accept saga options and start a watcher', () => {
      // arrange
      const expectedResult = fork(watchFetchSaga, payload);
      // act
      const result = Generator.next().value;
      // assert
      expect(result).toEqual(expectedResult);
    });
    it('should trigger the watcher once', () => {
      // arrange
      const expectedResult = put({ type: payload.requestId });
      // act
      const result = Generator.next(watcher).value;
      // assert
      expect(result).toEqual(expectedResult);
    });
    it('should listen for a LOCATION_CHANGE action and cancel the watcher', () => {
      // arrange
      let expectedResult = take(LOCATION_CHANGE);
      // act
      let result = Generator.next().value;
      // assert
      expect(result).toEqual(expectedResult);
      // arrange
      expectedResult = cancel(watcher);
      // act
      result = Generator.next().value;
      // assert
      expect(result).toEqual(expectedResult);
    });
  });

  describe('watchFetchSaga', () => {
    it('should start a watcher for the options.requestId', () => {
      // arrange
      const Generator = watchFetchSaga(payload);
      const expectedResult = fork(takeLatest, payload.requestId, callFetchSaga, payload);
      // act
      const result = Generator.next().value;
      // assert
      expect(result).toEqual(expectedResult);
    });
  });

  describe('callFetchSaga', () => {
    // we have three paths
    // with/without parentId, success, error
    describe('no parentId', () => {
      const { requestId, fetchGenerator } = payload;
      const Generator = callFetchSaga({ requestId, fetchGenerator });
      it('should dispatch asyncRequest action for the requestId', () => {
        // arrange
        const expectedResult = put(asyncRequest(requestId));
        // act
        const result = Generator.next().value;
        // assert
        expect(result).toEqual(expectedResult);
      });
    });

    describe('with parentId success', () => {
      const Generator = callFetchSaga(payload);
      const { requestId, fetchGenerator, parentId, resolveSelector } = payload;
      it('should wait for parent to resolve if parentId is passed', () => {
        // arrange
        const expectedResult = call(waitForOther, parentId, resolveSelector);
        // act
        const result = Generator.next().value;
        // assert
        expect(result).toEqual(expectedResult);
      });
      it('should dispatch asyncRequest action for the requestId', () => {
        // arrange
        const expectedResult = put(asyncRequest(requestId));
        // act
        const result = Generator.next().value;
        // assert
        expect(result).toEqual(expectedResult);
      });
      it('should call the requested fetchGenerator and wait to resolve', () => {
        // arrange
        const expectedResult = call(fetchGenerator, requestId);
        // act
        const result = Generator.next().value;
        // assert
        expect(result).toEqual(expectedResult);
      });
      it('should dispatch asyncSuccess action for the requestId', () => {
        // arrange
        const expectedResult = put(asyncSuccess(requestId));
        // act
        const result = Generator.next('responseValue').value;
        // assert
        expect(result).toEqual(expectedResult);
      });
    });

    describe('with error', () => {
      const Generator = callFetchSaga(payload);
      const { requestId } = payload;
      const error = new Error('Lex');
      it('should dispatch asyncRequest action for the requestId', () => {
        // arrange
        const expectedResult = put(asyncFail(requestId, error));
        // act
        Generator.next(); // wait for parent
        Generator.next(); // put request
        Generator.next(); // call generator
        const result = Generator.throw(error).value;
        // assert
        expect(result).toEqual(expectedResult);
      });
    });
  });
});
