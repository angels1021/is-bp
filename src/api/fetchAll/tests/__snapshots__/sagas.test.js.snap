// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API fetchAll sagas fetchFlow should error if options.requestId or options.fetchGenerator were passed 1`] = `"received is not a function"`;
